In [12]:
import pandas as pd


credit_card_df = pd.read_csv('Credit_card.csv')
credit_card_label_df = pd.read_csv('Credit_card_label.csv')


credit_card_df.head()
credit_card_label_df.head()


credits_df = pd.merge(credit_card_df, credit_card_label_df, on='Ind_ID')


credits_df.head()

credits_df.to_csv('merged_credit_data.csv', index=False)

credits_df


Ind_ID GENDER Car_Owner Propert_Owner  CHILDREN  Annual_income  \
0     5008827      M         Y             Y         0       180000.0   
1     5009744      F         Y             N         0       315000.0   
2     5009746      F         Y             N         0       315000.0   
3     5009749      F         Y             N         0            NaN   
4     5009752      F         Y             N         0       315000.0   
...       ...    ...       ...           ...       ...            ...   
1543  5028645      F         N             Y         0            NaN   
1544  5023655      F         N             N         0       225000.0   
1545  5115992      M         Y             Y         2       180000.0   
1546  5118219      M         Y             N         0       270000.0   
1547  5053790      F         Y             Y         0       225000.0   

               Type_Income                      EDUCATION  \
0                Pensioner               Higher education   
1     Commercial associate               Higher education   
2     Commercial associate               Higher education   
3     Commercial associate               Higher education   
4     Commercial associate               Higher education   
...                    ...                            ...   
1543  Commercial associate               Higher education   
1544  Commercial associate              Incomplete higher   
1545               Working               Higher education   
1546               Working  Secondary / secondary special   
1547               Working               Higher education   

            Marital_status       Housing_type  Birthday_count  Employed_days  \
0                  Married  House / apartment        -18772.0         365243   
1                  Married  House / apartment        -13557.0           -586   
2                  Married  House / apartment             NaN           -586   
3                  Married  House / apartment        -13557.0           -586   
4                  Married  House / apartment        -13557.0           -586   
...                    ...                ...             ...            ...   
1543               Married  House / apartment        -11957.0          -2182   
1544  Single / not married  House / apartment        -10229.0          -1209   
1545               Married  House / apartment        -13174.0          -2477   
1546        Civil marriage  House / apartment        -15292.0           -645   
1547               Married  House / apartment        -16601.0          -2859   

      Mobile_phone  Work_Phone  Phone  EMAIL_ID Type_Occupation  \
0                1           0      0         0             NaN   
1                1           1      1         0             NaN   
2                1           1      1         0             NaN   
3                1           1      1         0             NaN   
4                1           1      1         0             NaN   
...            ...         ...    ...       ...             ...   
1543             1           0      0         0        Managers   
1544             1           0      0         0     Accountants   
1545             1           0      0         0        Managers   
1546             1           1      1         0         Drivers   
1547             1           0      0         0             NaN   

      Family_Members  label  
0                  2      1  
1                  2      1  
2                  2      1  
3                  2      1  
4                  2      1  
...              ...    ...  
1543               2      0  
1544               1      0  
1545               4      0  
1546               2      0  
1547               2      0  

[1548 rows x 19 columns]

# (5 pts) Clean your dataset to turn categorical values into numerical ones. One-hot encoding is likely the answer, but it depends on the dataset. Your data may have ordinal columns, for example where one-hot encoding is not as appropriate.

In [14]:
credits_df.drop(columns=['Ind_ID'], inplace= True) 
from sklearn.preprocessing import LabelEncoder 

label_encoder = LabelEncoder()
credits_df['GENDER'] = label_encoder.fit_transform(credits_df['GENDER']) 
credits_df['Car_Owner'] = label_encoder.fit_transform(credits_df['Car_Owner'])
credits_df['Propert_Owner'] = label_encoder.fit_transform(credits_df['Propert_Owner'])
credits_df['Type_Income'] = label_encoder.fit_transform(credits_df['Type_Income'])
credits_df['EDUCATION'] = label_encoder.fit_transform(credits_df['EDUCATION'])
credits_df['Marital_status'] = label_encoder.fit_transform(credits_df['Marital_status'])
credits_df['Housing_type'] = label_encoder.fit_transform(credits_df['Housing_type'])
credits_df['Type_Occupation'] = label_encoder.fit_transform(credits_df['Type_Occupation'])

credits_df.head()

GENDER  Car_Owner  Propert_Owner  CHILDREN  Annual_income  Type_Income  \
0       1          1              1         0       180000.0            1   
1       0          1              0         0       315000.0            0   
2       0          1              0         0       315000.0            0   
3       0          1              0         0            NaN            0   
4       0          1              0         0       315000.0            0   

   EDUCATION  Marital_status  Housing_type  Birthday_count  Employed_days  \
0          1               1             1        -18772.0         365243   
1          1               1             1        -13557.0           -586   
2          1               1             1             NaN           -586   
3          1               1             1        -13557.0           -586   
4          1               1             1        -13557.0           -586   

   Mobile_phone  Work_Phone  Phone  EMAIL_ID  Type_Occupation  Family_Members  \
0             1           0      0         0               18               2   
1             1           1      1         0               18               2   
2             1           1      1         0               18               2   
3             1           1      1         0               18               2   
4             1           1      1         0               18               2   

   label  
0      1  
1      1  
2      1  
3      1  
4      1

# (3 pts) Perform univariate linear regression on the dataset. Select your variable to predict. How well did this model perform? Is this a good approach for this dataset? Why or why not? 

In [24]:
from sklearn.preprocessing import MinMaxScaler

def perform_normalization(df):
   
    numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns

    scaler = MinMaxScaler()

    df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

    return df

normalized_df = perform_normalization(credits_df)

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

predictor_variable = 'Annual_income'
target_variable = 'label'

filtered_data = credits_df[[predictor_variable, target_variable]].dropna()

X = filtered_data[[predictor_variable]]
y = filtered_data[target_variable]

model = LinearRegression()

model.fit(X, y)

y_pred = model.predict(X)

mse = mean_squared_error(y, y_pred)
r2 = r2_score(y, y_pred)

print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)
print("Mean Squared Error:", mse)
print("R^2 Score:", r2)


Coefficients: [0.11671915]
Intercept: 0.09756938365879686
Mean Squared Error: 0.09744264052629872
R^2 Score: 0.0007538345666244783


This is not the best model to use when trying to find credit approval predictions. The r2 score is very low. when the r2 score is closer to 0 then that means the prediction is farther off from the actual value. When the r2 score is closer to 1 then that means the prediction is closer to the actual value. In this case, the r2 score is low which means that this is not the best model to find the credit card approval predictions. 

# (8 pts) Perform KNN on this dataset. As part of this, write a function that selects the optimal value of k. How well did this model perform?

In [22]:
from sklearn.preprocessing import MinMaxScaler

def perform_normalization(df):
    numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns

    scaler = MinMaxScaler()

    df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

    return df


normalized_df = perform_normalization(credits_df)


from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

def select_optimal_k(X_train, y_train, X_test, y_test, max_k=10):
    best_k = 0
    best_accuracy = 0

    for k in range(1, max_k + 1):
       
        knn = KNeighborsClassifier(n_neighbors=k)
        
        knn.fit(X_train, y_train)
        
        y_pred = knn.predict(X_test)
        
        accuracy = accuracy_score(y_test, y_pred)
        
        if accuracy > best_accuracy:
            best_k = k
            best_accuracy = accuracy
    
    return best_k, best_accuracy

predictor_variables = ['Annual_income']
target_variable = 'label'

filtered_data = credits_df[predictor_variables + [target_variable]].dropna()

X = filtered_data[predictor_variables]
y = filtered_data[target_variable]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

optimal_k, accuracy = select_optimal_k(X_train, y_train, X_test, y_test, max_k=10)
print("Optimal value of k:", optimal_k)
print("Accuracy:", accuracy)


Optimal value of k: 4
Accuracy: 0.9114754098360656


The KNN model worked relatively well based on looking at the accuracy. The accuracy rating is pretty high which tells me that the model did a good job of predicting the credit card approval for the dataframe. 

# (6 pts) Work with your dataset to perform logistic regression. How well did this perform

In [25]:
from sklearn.preprocessing import MinMaxScaler

def perform_normalization(df):
    numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns

    scaler = MinMaxScaler()

    df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

    return df
    
normalized_df = perform_normalization(credits_df)


from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

def perform_logistic_regression(df, target_variable):
    X = df.drop(columns=[target_variable])
    y = df[target_variable]

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)

    logistic_model = LogisticRegression(max_iter=1000)  # Increase max_iter
    logistic_model.fit(X_train_scaled, y_train)

    y_pred = logistic_model.predict(X_test_scaled)

    accuracy = accuracy_score(y_test, y_pred)

    return {'Accuracy': accuracy}

results = perform_logistic_regression(credits_df_cleaned, 'label')
print(results)


{'Accuracy': 0.9036544850498339}


This model works pretty well. With an accuracy rating of 0.9036... it means that the model predicts whether or not someone got approved for a credit card correctly 90.4% percent of the time. Based on the accuracy score, the KNN model still works better than this one. 

# (3 pts) Perform normalization on your dataset. Does it change the performance for 2-4? What is the best measure of performance for your dataset (accuracy or something else) and why?


I went ahead and performed normalization on the other questions. It didn't look like it changed the performance of the datasets and my results still look the same. I think that the best measure of performance for my datasets is accuracy because that gives me a straightforward idea of how accurate my model is at predicting the relationship between my dependent and independent variables. 